pkg/api/steve/collections/rbac.go

```diff
@@ -0,0 +1,315 @@
+	}
+
+	subjects := rb.Subjects
+	roleRef := rb.RoleRef
+	roleRef.Name += suffix
+
+	collectionRoleBindingName := pkgrbac.NameForRoleBinding(namespace.Name, roleRef, subjects[0])
+	collectionRoleBinding, err := h.roleBindingsCache.Get(namespace.Name, collectionRoleBindingName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return rb, nil
+	}
+	if collectionRoleBinding == nil {
+		collectionRoleBinding, err = h.createRoleBinding(collectionRoleBindingName, namespace.Name, roleRef, subjects, rb)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if collectionRoleBinding.RoleRef.Kind == "ClusterRole" {
+		collectionClusterRole, err := h.clusterRolesCache.Get(collectionRoleBinding.RoleRef.Name)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		if collectionClusterRole == nil {
+			err = h.createClusterRole(collectionRoleBinding.RoleRef.Name, collectionRules, clusterRole)
+			if err != nil {
+				return nil, err
+			}
+			return rb, nil
+		}
+		if _, ok := collectionClusterRole.GetAnnotations()[clusterRoleOwnerAnnotation]; ok { // owned by roletemplate controllers, don't update it
+			return rb, nil
+		}
+		err = h.compareAndUpdateClusterRole(collectionClusterRole, collectionRules, clusterRole)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	collectionRole, err := h.rolesCache.Get(namespace.Name, collectionRoleBinding.RoleRef.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	if collectionRole == nil {
+		err = h.createRole(collectionRoleBinding.RoleRef.Name, namespace.Name, collectionRules, role)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	err = h.compareAndUpdateRole(collectionRole, collectionRules, role)
+	if err != nil {
+		return nil, err
+	}
+
+	return rb, nil
+}
+
+func GetCollectionRules(apis APIResourceWatcher, rules []rbacv1.PolicyRule) ([]rbacv1.PolicyRule, error) {
+	collectionRules := []rbacv1.PolicyRule{}
+	validRules := []rbacv1.PolicyRule{}
+	for _, rule := range rules {
+		if len(rule.ResourceNames) != 0 {
+			continue
+		}
+		if len(rule.NonResourceURLs) != 0 {
+			continue
+		}
+		for _, verb := range rule.Verbs {
+			if verb == "*" || verb == "list" {
+				validRules = append(validRules, rule)
+			}
+		}
+	}
+	if len(validRules) == 0 {
+		return collectionRules, nil
+	}
+	if len(apis.List()) == 0 {
+		return collectionRules, ErrNotReady
+	}
+	for _, rule := range validRules {
+		for _, apiGroup := range rule.APIGroups {
+			for _, resource := range rule.Resources {
+				_, ok := apis.Get(resource, apiGroup)
+				if ok {
+					if len(collectionRules) == 0 {
+						collectionRules = []rbacv1.PolicyRule{
+							{
+								APIGroups: []string{Group},
+								Verbs:     []string{"list"},
+							},
+						}
+					}
+					if apiGroup != "" {
+						resource = apiGroup + "." + resource
+					}
+					collectionRules[0].Resources = append(collectionRules[0].Resources, resource)
+					continue
+				}
+				if resource == "*" {
```

Comment:
Nit: Since this represents the special case (all resources in group), I'd check for this first and then do the check on the non-* use case.

Reply:
The non-special case is shorter and reads easier to me being up front :woman_shrugging:

Reply:
Ok - resolving since this is based on subjective tastes and you are the PR author.